util/webhook/webhook.go

```diff
@@ -288,27 +306,54 @@ func (a *ArgoCDWebhookHandler) HandleEvent(payload interface{}) {
 			log.Warnf("Failed to get repoRegexp: %s", err)
 			continue
 		}
-		for _, app := range filteredApps {
-
-			for _, source := range app.Spec.GetSources() {
-				if sourceRevisionHasChanged(source, revision, touchedHead) && sourceUsesURL(source, webURL, repoRegexp) {
-					if appFilesHaveChanged(&app, changedFiles) {
-						namespacedAppInterface := a.appClientset.ArgoprojV1alpha1().Applications(app.ObjectMeta.Namespace)
-						_, err = argo.RefreshApp(namespacedAppInterface, app.ObjectMeta.Name, v1alpha1.RefreshTypeNormal)
-						if err != nil {
-							log.Warnf("Failed to refresh app '%s' for controller reprocessing: %v", app.ObjectMeta.Name, err)
-							continue
-						}
-						// No need to refresh multiple times if multiple sources match.
-						break
-					} else if change.shaBefore != "" && change.shaAfter != "" {
-						if err := a.storePreviouslyCachedManifests(&app, change, trackingMethod, appInstanceLabelKey); err != nil {
-							log.Warnf("Failed to store cached manifests of previous revision for app '%s': %v", app.Name, err)
-						}
-					}
+
+		maxConcurrentAppRefresh := a.maxConcurrentAppRefresh
+		appBacklog := make(chan v1alpha1.Application, len(filteredApps))
+
+		for _, filteredApp := range filteredApps {
+			appBacklog <- filteredApp
+		}
+		close(appBacklog)
+
+		var wg sync.WaitGroup
+		wg.Add(maxConcurrentAppRefresh)
+
+		for i := 0; i < maxConcurrentAppRefresh; i++ {
+			go func() {
+				defer wg.Done()
+				for app := range appBacklog {
+					a.refreshChangedAppOrStoreCachedManifests(app, revision, touchedHead, webURL, repoRegexp, changedFiles, change, trackingMethod, appInstanceLabelKey)
 				}
+			}()
+		}
+		wg.Wait()
+	}
+}
+
+func (a *ArgoCDWebhookHandler) refreshChangedAppOrStoreCachedManifests(app v1alpha1.Application, revision string, touchedHead bool, webURL string,
+	repoRegexp *regexp.Regexp, changedFiles []string, change changeInfo, trackingMethod string, appInstanceLabelKey string) {
+	for _, source := range app.Spec.GetSources() {
+		if !(sourceRevisionHasChanged(source, revision, touchedHead) && sourceUsesURL(source, webURL, repoRegexp)) {
+			continue
+		}
+		if !appFilesHaveChanged(&app, changedFiles) {
+			// we cannot retrieve previous and next commit SHA for bitbucket.
```

Comment:
added this comment based on ref:
https://github.com/argoproj/argo-cd/pull/4699/files#r514417202